mne/beamformer/_ap.py

```diff
@@ -0,0 +1,712 @@
+    gain : array, shape (nchannels, ndipoles)
+        Gain matrix.
+
+    Returns
+    -------
+    s_ap : list of int
+        List of dipole indices.
+
+    """
+    s_ap = []
+    ap_val1 = np.zeros(attr_dict['ndipoles'])
+    for dip in range(attr_dict['ndipoles']):
+        l_p = gain[:, [dip]]
+        ap_val1[dip] = multi_dot([l_p.T, data_cov, l_p]) / (l_p.T @ l_p)[0, 0]
+    s1_idx = np.argmax(ap_val1)
+    s_ap.append(s1_idx)
+    return s_ap
+
+
+def _fixed_phase1b(gain, s_ap, data_cov, attr_dict):
+    """Calculate phase 1b of fixed oriented AP.
+
+    Adding one source at a time.
+
+    Parameters
+    ----------
+    gain : array, shape (nchannels, ndipoles)
+        Gain matrix.
+    s_ap : list of int
+        List of dipole indices.
+    data_cov : array
+        Data Covariance.
+    attr_dict : dict
+        See: _calculate_fixed_alternating_projections.
+
+    Returns
+    -------
+    s_ap : list of int
+        List of dipole indices.
+
+    """
+    for _ in range(1, attr_dict['nsources']):
+        ap_val2 = np.zeros(attr_dict['ndipoles'])
+        sub_g = gain[:, s_ap]
+        act_spc = _active_subspace(sub_g)
+        perpend_spc = np.eye(act_spc.shape[0]) - act_spc
+        for dip in range(attr_dict['ndipoles']):
+            l_p = np.expand_dims(gain[:, dip], axis=1)
+            ap_val2[dip] = multi_dot(
+                [l_p.T, perpend_spc, data_cov, perpend_spc, l_p]
+            ) / ((multi_dot([l_p.T, perpend_spc, l_p]))[0, 0])
+        s2_idx = np.argmax(ap_val2)
+        s_ap.append(s2_idx)
+    logger.info('current s_ap = {}'.format(s_ap))
+    return s_ap
+
+
+def _fixed_phase2(attr_dict, s_ap_2, gain, data_cov):
+    """Calculate phase 2 of fixed oriented AP.
+
+    Altering the projection of current estimated dipoles.
+
+    Parameters
+    ----------
+    attr_dict : dict
+        See: _calculate_fixed_alternating_projections.
+    s_ap_2 : list of int
+        List of dipole indices.
+    gain : array, shape (nchannels, ndipoles)
+        Gain matrix.
+    data_cov : array
+        Data Covariance.
+
+    Returns
+    -------
+    s_ap_2 : list of int
+        List of dipole indices.
+
+    """
+    for itr in range(attr_dict['max_iter']):
+        logger.info('iteration No. {}'.format(itr + 1))
+        s_ap_2_prev = copy(s_ap_2)
+        for src in range(attr_dict['nsources']):
+            # AP localization of src-th source
+            ap_val2 = np.zeros(attr_dict['ndipoles'])
+            s_ap_temp = copy(s_ap_2)
+            s_ap_temp.pop(src)
+            sub_g = gain[:, s_ap_temp]
+            act_spc = _active_subspace(sub_g)
+            perpend_spc = np.eye(act_spc.shape[0]) - act_spc
+            for dip in range(attr_dict['ndipoles']):
+                l_p = np.expand_dims(gain[:, dip], axis=1)
+                ap_val2[dip] = multi_dot(
+                    [l_p.T, perpend_spc, data_cov, perpend_spc, l_p]
+                ) / ((multi_dot([l_p.T, perpend_spc, l_p]))[0, 0])
```

Comment:
idem. The for loop can be potentially avoided

Reply:
also if you can point to equation numbers in the paper that would also help reviewing this code. thx

Reply:
In [AdlerEtAl2022](https://arxiv.org/abs/2202.01120v1), Eq. (22) [page 3] deals with this exact loop, while Eq. (21) describes _fixed_phase1a().